File: bettervideoplayer/src/main/java/com/halilibo/bettervideoplayer/OnSwipeTouchListener.java
package com.halilibo.bettervideoplayer;

import android.os.Handler;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;

/**
 * @author: Halil Ozercan
 */
abstract class OnSwipeTouchListener implements OnTouchListener {

    private final boolean doubleTapEnabled;
    private final Handler mHandler;
    private Runnable futureClickRunnable = new Runnable() {
        @Override
        public void run() {
            onClick();
        }
    };

    enum Direction{
        LEFT, RIGHT, UP, DOWN;
    }
    public OnSwipeTouchListener(boolean doubleTapEnabled){
        this.doubleTapEnabled = doubleTapEnabled;
        this.mHandler = new Handler();
    }

    private final static String TAG = "ClickFrame";
    private final static int SWIPE_THRESHOLD = 100;
    private static final long DOUBLE_TAP_THRESHOLD = 150;

    // 0: uninitialized
    // 1: horizontal
    // 2: vertical
    private int initialGesture;

    protected float initialX;
    protected float initialY;
    private float decidedX;
    private float decidedY;
    private long lastClick = 0;
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        int action = event.getActionMasked();
        switch (action) {

            case MotionEvent.ACTION_DOWN:
                initialX = event.getX();
                initialY = event.getY();
                initialGesture = 0;
                break;

            case MotionEvent.ACTION_MOVE:
                float deltaX;
                float deltaY;
                if(initialGesture == 0) {
                    deltaX = event.getX() - initialX;
                    deltaY = event.getY() - initialY;
                }
                else{
                    deltaX = event.getX() - decidedX;
                    deltaY = event.getY() - decidedY;
                }

                if(initialGesture == 0 && Math.abs(deltaX) > SWIPE_THRESHOLD){
                    initialGesture = 1;
                    decidedX = event.getX();
                    decidedY = event.getY();
                    if(deltaX > 0){
                        onBeforeMove(Direction.RIGHT);
                    }
                    else{
                        onBeforeMove(Direction.LEFT);
                    }
                }
                else if(initialGesture == 0 && Math.abs(deltaY) > SWIPE_THRESHOLD){
                    initialGesture = 2;
                    decidedX = event.getX();
                    decidedY = event.getY();
                    if(deltaY > 0){
                        onBeforeMove(Direction.DOWN);
                    }
                    else{
                        onBeforeMove(Direction.UP);
                    }
                }

                if(initialGesture == 1){
                    if(deltaX > 0){
                        onMove(Direction.RIGHT, deltaX);
                    }
                    else{
                        onMove(Direction.LEFT, -deltaX);
                    }
                }
                else if(initialGesture == 2){
                    if(deltaY > 0){
                        onMove(Direction.DOWN, deltaY);
                    }
                    else{
                        onMove(Direction.UP, -deltaY);
                    }
                }

                break;

            case MotionEvent.ACTION_UP:
                if(initialGesture == 0){ // Finger did not move enough to trigger a swipe
                    if(doubleTapEnabled &&
                        System.currentTimeMillis() - lastClick <= DOUBLE_TAP_THRESHOLD &&
                        lastClick != 0) {
                        mHandler.removeCallbacks(futureClickRunnable);
                        onDoubleTap(event);
                        return true;
                    }
                    else {
                        lastClick = System.currentTimeMillis();
                        if(doubleTapEnabled)
                            mHandler.postDelayed(futureClickRunnable, DOUBLE_TAP_THRESHOLD);
                        else
                            mHandler.post(futureClickRunnable);
                        return true;
                    }
                }

                onAfterMove();
                initialGesture = 0;
                return true;

            case MotionEvent.ACTION_CANCEL:
                break;

            case MotionEvent.ACTION_OUTSIDE:
                break;
        }

        return true;
    }

    public abstract void onMove(Direction dir, float diff);

    public abstract void onClick();

    public abstract void onDoubleTap(MotionEvent event);

    public abstract void onAfterMove();

    public abstract void onBeforeMove(Direction dir);
}

File: bvpkotlin/src/main/java/com/halilibo/bvpkotlin/OnSwipeTouchListener.kt
package com.halilibo.bvpkotlin

import android.os.Handler
import android.view.MotionEvent
import android.view.View

/**
 * @author: Halil Ozercan
 */
internal abstract class OnSwipeTouchListener(
        private val doubleTapEnabled: Boolean
) : View.OnTouchListener {

    private val mHandler: Handler = Handler()
    private val futureClickRunnable = Runnable { onClick() }

    // 0: uninitialized
    // 1: horizontal
    // 2: vertical
    private var initialGesture: Int = 0

    protected var initialX: Float = 0.toFloat()
    protected var initialY: Float = 0.toFloat()
    private var decidedX: Float = 0.toFloat()
    private var decidedY: Float = 0.toFloat()
    private var lastClick: Long = 0

    internal enum class Direction {
        LEFT, RIGHT, UP, DOWN
    }

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        val action = event.actionMasked
        when (action) {

            MotionEvent.ACTION_DOWN -> {
                initialX = event.x
                initialY = event.y
                initialGesture = 0
            }

            MotionEvent.ACTION_MOVE -> {
                val deltaX: Float
                val deltaY: Float
                if (initialGesture == 0) {
                    deltaX = event.x - initialX
                    deltaY = event.y - initialY
                } else {
                    deltaX = event.x - decidedX
                    deltaY = event.y - decidedY
                }

                if (initialGesture == 0 && Math.abs(deltaX) > SWIPE_THRESHOLD) {
                    initialGesture = 1
                    decidedX = event.x
                    decidedY = event.y
                    if (deltaX > 0) {
                        onBeforeMove(Direction.RIGHT)
                    } else {
                        onBeforeMove(Direction.LEFT)
                    }
                } else if (initialGesture == 0 && Math.abs(deltaY) > SWIPE_THRESHOLD) {
                    initialGesture = 2
                    decidedX = event.x
                    decidedY = event.y
                    if (deltaY > 0) {
                        onBeforeMove(Direction.DOWN)
                    } else {
                        onBeforeMove(Direction.UP)
                    }
                }

                if (initialGesture == 1) {
                    if (deltaX > 0) {
                        onMove(Direction.RIGHT, deltaX)
                    } else {
                        onMove(Direction.LEFT, -deltaX)
                    }
                } else if (initialGesture == 2) {
                    if (deltaY > 0) {
                        onMove(Direction.DOWN, deltaY)
                    } else {
                        onMove(Direction.UP, -deltaY)
                    }
                }
            }

            MotionEvent.ACTION_UP -> {
                if (initialGesture == 0) { // Finger did not move enough to trigger a swipe
                    return if (doubleTapEnabled &&
                            System.currentTimeMillis() - lastClick <= DOUBLE_TAP_THRESHOLD &&
                            lastClick != 0L) {
                        mHandler.removeCallbacks(futureClickRunnable)
                        onDoubleTap(event)
                        true
                    } else {
                        lastClick = System.currentTimeMillis()
                        if (doubleTapEnabled)
                            mHandler.postDelayed(futureClickRunnable, DOUBLE_TAP_THRESHOLD)
                        else
                            mHandler.post(futureClickRunnable)
                        true
                    }
                }

                onAfterMove()
                initialGesture = 0
                return true
            }

            MotionEvent.ACTION_CANCEL -> {
            }

            MotionEvent.ACTION_OUTSIDE -> {
            }
        }

        return true
    }

    abstract fun onMove(dir: Direction, diff: Float)

    abstract fun onClick()

    abstract fun onDoubleTap(event: MotionEvent)

    abstract fun onAfterMove()

    abstract fun onBeforeMove(dir: Direction)

    companion object {

        private const val SWIPE_THRESHOLD = 100
        private const val DOUBLE_TAP_THRESHOLD: Long = 150
    }
}